src/pudl/etl/check_foreign_keys.py

```diff
@@ -50,15 +61,154 @@ def pudl_check_fks(logfile: pathlib.Path, loglevel: str):
+            return (
+                (self.child_table == other.child_table)
+                and (self.parent_table == other.parent_table)
+                and (self.foreign_key == other.foreign_key)
+                and (self.rowids == other.rowids)
+            )
+        return False
+
+
+class ForeignKeyErrors(sa.exc.SQLAlchemyError):  # noqa: N818
+    """Raised when data in a database violate multiple foreign key constraints."""
+
+    def __init__(self, fk_errors: list[ForeignKeyError]):
+        """Initialize a new ForeignKeyErrors object."""
+        self.fk_errors = fk_errors
+
+    def __str__(self):
+        """Create string representation of ForeignKeyErrors object."""
+        fk_errors = [str(x) for x in self.fk_errors]
+        return "\n".join(fk_errors)
+
+    def __iter__(self):
+        """Iterate over the fk errors."""
+        return self.fk_errors
+
+    def __getitem__(self, idx):
+        """Index the fk errors."""
+        return self.fk_errors[idx]
+
+
+def _get_fk_list(engine: sa.Engine, table: str) -> pd.DataFrame:
+    """Retrieve a dataframe of foreign keys for a table.
+
+    Description from the SQLite Docs: 'This pragma returns one row for each foreign
+    key constraint created by a REFERENCES clause in the CREATE TABLE statement of
+    table "table-name".'
+
+    The PRAGMA returns one row for each field in a foreign key constraint. This
+    method collapses foreign keys with multiple fields into one record for
+    readability.
+    """
+    with engine.begin() as con:
+        table_fks = pd.read_sql_query(f"PRAGMA foreign_key_list({table});", con)
+
+    # Foreign keys with multiple fields are reported in separate records.
+    # Combine the multiple fields into one string for readability.
+    # Drop duplicates so we have one FK for each table and foreign key id
+    table_fks["fk"] = table_fks.groupby("table")["to"].transform(
+        lambda field: "(" + ", ".join(field) + ")"
+    )
+    table_fks = table_fks[["id", "table", "fk"]].drop_duplicates()
+
+    # Rename the fields so we can easily merge with the foreign key errors.
+    table_fks = table_fks.rename(columns={"id": "fkid", "table": "parent"})
+    table_fks["table"] = table
+    return table_fks
+
+
+def check_foreign_keys(engine: sa.Engine):
+    """Check foreign key relationships in the database.
+
+    The order assets are loaded into the database will not satisfy foreign key
+    constraints so we can't enable foreign key constraints. However, we can
+    check for foreign key failures once all of the data has been loaded into
+    the database using the `foreign_key_check` and `foreign_key_list` PRAGMAs.
+
+    You can learn more about the PRAGMAs in the `SQLite docs
+    <https://www.sqlite.org/pragma.html#pragma_foreign_key_check>`__.
+
+    Raises:
+        ForeignKeyErrors: if data in the database violate foreign key constraints.
+    """
+    logger.info(f"Running foreign key check on {engine.name} database.")
```

Comment:
I think this is the logging message you're referring to above, but unlike the logging message which was removed, it doesn't provide the path to the database whose FK relationships it's checking, which could lead to confusion in testing or development if say, the `$PUDL_OUTPUT` envvar isn't set correctly, or isn't set to the value that the user thinks it's set to.